test/unit/lexer/location_offset.d

```diff
@@ -0,0 +1,489 @@
+    "for_" : Test("for"),
+    "do_" : Test("do"),
+    "switch_" : Test("switch"),
+    "case_" : Test("case"),
+    "default_" : Test("default"),
+    "break_" : Test("break"),
+    "continue_" : Test("continue"),
+    "with_" : Test("with"),
+    "synchronized_" : Test("synchronized"),
+    "return_" : Test("return"),
+    "goto_" : Test("goto"),
+    "try_" : Test("try"),
+    "catch_" : Test("catch"),
+    "finally_" : Test("finally"),
+    "asm_" : Test("asm"),
+    "foreach_" : Test("foreach"),
+    "foreach_reverse_" : Test("foreach_reverse"),
+    "scope_" : Test("scope"),
+
+    "invariant_" : Test("invariant"),
+
+    "unittest_" : Test("unittest"),
+
+    "argumentTypes" : Test("__argTypes"),
+    "ref_" : Test("ref"),
+    "macro_" : Test("macro"),
+
+    "parameters" : Test("__parameters"),
+    "traits" : Test("__traits"),
+    "overloadSet" : Test("__overloadset"),
+    "pure_" : Test("pure"),
+    "nothrow_" : Test("nothrow"),
+    "gshared" : Test("__gshared"),
+
+    "line" : Test("__LINE__"),
+    "file" : Test("__FILE__"),
+    "fileFullPath" : Test("__FILE_FULL_PATH__"),
+    "moduleString" : Test("__MODULE__"),
+    "functionString" : Test("__FUNCTION__"),
+    "prettyFunction" : Test("__PRETTY_FUNCTION__"),
+
+    "shared_" : Test("shared"),
+    "at" : Test("at sign", "@"),
+    "pow" : Test("power", "^^"),
+    "powAssign" : Test("power assign", "^^="),
+    "goesTo" : Test("fat arrow", "=>"),
+    "vector" : Test("__vector"),
+    "pound" : Test("pound", "#"),
+];
+
+// Ignore tokens not produced by the lexer or tested above
+enum ignoreTokens
+{
+    reserved,
+    negate,
+    array,
+    call,
+    address,
+    star,
+    type,
+    dotVariable,
+    dotIdentifier,
+    dotTemplateInstance,
+    dotType,
+    symbolOffset,
+    variable,
+    arrayLength,
+    dotTemplateDeclaration,
+    declaration,
+    dSymbol,
+    uadd,
+    remove,
+    newAnonymousClass,
+    comment,
+    arrayLiteral,
+    assocArrayLiteral,
+    structLiteral,
+    classReference,
+    thrownException,
+    delegatePointer,
+    delegateFunctionPointer,
+    identity,
+    notIdentity,
+    index,
+    concatenate,
+    concatenateElemAssign,
+    concatenateDcharAssign,
+    construct,
+    blit,
+    arrow,
+    prePlusPlus,
+    preMinusMinus,
+    int128Literal,
+    uns128Literal,
+    halt,
+    tuple,
+    error,
+    int128,
+    uns128,
+    onScopeExit,
+    onScopeFailure,
+    onScopeSuccess,
+    interval,
+    voidExpression,
+    cantExpression,
+    showCtfeContext,
+    objcClassReference,
+    vectorArray,
+    max_,
+};
+
+static foreach (tok; __traits(allMembers, TOK))
+{
+    static if (!__traits(hasMember, ignoreTokens, tok))
+    {
+        @(tests[tok].description)
+        unittest
+        {
+            const newCode = "first_token " ~ tests[tok].code;
+
+            scope lexer = new Lexer("test.d", newCode.ptr, 0, newCode.length, 0, 0);
+
+            lexer.nextToken;
+            lexer.nextToken;
+
```

Comment:
A couple of times I have seen that the token and the source code doesn't match. You should probably add an extra assertion (yes, I know that goes agains what I said before about a single assertion per test) that verifies that the input token (that is, the key in the associative array) matches the output token.